### PR TITLE
Use `GenericForeignKey` and `GenericRelation` for `Revision`

### LINKED ISSUE #8570: Use `GenericForeignKey` and `GenericRevision` for `Revision` model
### Is your proposal related to a problem?

From #8441:

> ... We are not using the `GenericForeignKey` field due to complexities it adds when handling the default/specific Page models (see also [ticket 16055 in Django](https://code.djangoproject.com/ticket/16055)). However, we try to simulate the same behaviour as the `GenericForeignKey` wherever possible.

To summarise a previous discussion with @jacobtoppm and @kaedroho:

While this allowed us to make the `Revision` model generic without changing much queries in the codebase, this solution is not ideal. Making `content_object` a `@property` instead of a `GenericForeignKey` makes us lose the information of the field in `Model._meta.get_fields()`. It also removes the ability to use related queries (e.g. `revision__content_object` or `page__revisions`). Ideally, we should use `GenericForeignKey` and `GenericRelation` and fix/work around any issues with existing queries/tests that prevent us from doing so.

### Describe the solution you'd like

Use `GenericForeignKey` for `content_object` in the `Revision` model, and use `GenericRelation` in the `Page` model.

### Describe alternatives you've considered

Leave the code as it currently is (using custom-made `@property`).

### Additional context

Some of the changes to the tests in #7221 may be useful for this, as I encountered some failing tests around the base/specific Page model when initially trying to use `GenericForeignKey`/`GenericRelation`.

### PR DESCRIPTION
**Pull request by laymonage**
Fixes #8570, needs #8573 to be merged first.

_Please check the following:_

- [x] Do the tests still pass?[^1]
- [x] Does the code comply with the style guide? 
    - [x] Run `make lint` from the Wagtail root. 
- [ ] For Python changes: Have you added tests to cover the new/fixed behaviour?
- [ ] For front-end changes: Did you test on all of Wagtail’s supported environments?[^2]
    - [ ] **Please list the exact browser and operating system versions you tested**:
    - [ ] **Please list which assistive technologies [^3] you tested**: 
- [ ] For new features: Has the documentation been updated accordingly?

**Please describe additional details for testing this change**. 

[^1]: [Development Testing](https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
[^2]: [Browser and device support](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)
[^3]: [Accessibility Target](https://docs.wagtail.org/en/latest/contributing/developing.html#accessibility-targets)

### CODE REVIEW

File: wagtail/models/__init__.py [hunk 1]
```diff
@@ -2238,15 +2235,9 @@ class Revision(models.Model):
...
-        return self.base_content_type.get_object_for_this_type(pk=self.object_id)
-
-    @cached_property
-    def specific_content_object(self):
-        if isinstance(self.content_object, Page):
-            return self.content_object.specific
-        return self.content_type.get_object_for_this_type(pk=self.object_id)
+    content_object = GenericForeignKey(
+        "content_type", "object_id", for_concrete_model=False
+    )
```
Comment (laymonage) [inline wagtail/models/__init__.py:2240]: This means that the `content_object` is always an instance of the specific page type. I don't think there's a downside to this as all tests seem to pass (after some minor adjustments in this PR).

**Review by laymonage — COMMENTED** [review]
Heh, it seems that it's possible to use `GenericForeignKey` after all! @jacobtoppm @kaedroho

I guess the trick is to use the specific page model as the `content_type`. ~~but still using the base Page model in queries (by using `base_content_type` in the `page_revisions` manager). However, this means that we can't use `GenericRelation` in queries as it doesn't seem to return the revisions using the `content_type` for some reason... For the record: the tests pass without the last commit. (Could this be an issue with some of the tests fixtures?)~~

Edit: Just found the main issue: we have code that access `.revisions` from the base Page instance. This wasn't an issue when `PageRevision` was not generic. With `GenericRelation`, this is an issue as Django will look for revisions using the base Page model as the `content_type`, while we create them with the specific content type. I made a workaround using a property that returns `self.specific_deferred._revisions` where `_revisions` is the `GenericRelation`, and it seems to work, but we need to get #7221 rebased and merged first.

Meanwhile, [the issue with PostgreSQL](https://code.djangoproject.com/ticket/16055) only happens when using the `GenericRelation` in a query, which previously would happen if we do a `select_related()` (or rather, `prefetch_related()` as `select_related()` doesn't work with `GenericRelation`) of the `Page` from a `Revision` query. Since the `select_related()` for these cases have been removed in the previous PR, the error does not happen.

File: wagtail/models/__init__.py [hunk 2]
```diff
@@ -317,14 +318,15 @@ class Page(AbstractPage, index.Indexed, ClusterableModel, metaclass=PageBase):
...
-        "Revision",
+        "wagtailcore.Revision",
         related_name="+",
         verbose_name=_("live revision"),
         on_delete=models.SET_NULL,
         null=True,
         blank=True,
         editable=False,
     )
+    _revisions = GenericRelation("wagtailcore.Revision", related_query_name="page")
```
Comment (laymonage) [inline wagtail/models/__init__.py:329]: Should this use a non-"private" name?

File: docs/releases/4.0.md [hunk 3]
```diff
@@ -47,3 +47,21 @@ The `PageRevision` model has been replaced with a generic `Revision` model. If y
...
  * Queries for `Page` objects that use `PageRevision.page_id` should be updated to cast the `Revision.object_id` to an integer before using it in the query.
+ * Access to `PageRevision.page` should be updated to `Revision.content_object`.
+
+If you are maintaining a package across multiple Wagtail versions that includes a model with a `ForeignKey` to the `PageRevision` model, you can create a helper function to correctly resolve the model depending on the installed Wagtail version, for example:
+
+```python
+from django.db import models
+from wagtail import VERSION as WAGTAIL_VERSION
+
+def get_revision_model():
```
Comment (zerolab) [inline docs/releases/4.0.md:58]: nitpick: always lint your code examples. this and the class definition need another blank line

File: wagtail/models/__init__.py (hunk 2 above)
Comment (kaedroho) [reply wagtail/models/__init__.py:329]: This looks OK to me, I don't think we'd ever want someone to call this directly?

**Review by kaedroho — APPROVED** [review]
Small change but a big improvement, nice work!

**Pull request merged**